hedera-node/hedera-app/src/main/java/com/hedera/node/app/tss/TssBaseServiceImpl.java

```diff
@@ -380,12 +407,359 @@ public TssMessage getTssMessageFromBytes(Bytes wrap, TssParticipantDirectory dir
+        if (numEncryptionKeys != targetRoster.rosterEntries().size()) {
+            return WAITING_FOR_ENCRYPTION_KEYS;
+        }
+        // Since this is called only once when JVM restarts, it is okay to do these synchronously.
+        final var activeDirectory = tssDirectoryAccessor.activeParticipantDirectory();
+        final var tssMessages = tssStore.getMessagesForTarget(targetRosterHash);
+        final var result = voteForValidMessages(tssMessages, activeDirectory, tssLibrary);
+        if (result.isEmpty()) {
+            return WAITING_FOR_THRESHOLD_TSS_MESSAGES;
+        } else {
+            return WAITING_FOR_THRESHOLD_TSS_VOTES;
+        }
+    }
+
+    /**
+     * Computes the next TSS status from the state.
+     *
+     * @param isStakePeriodBoundary whether the current consensus round is a stake period boundary
+     * @param consensusNow          the current consensus time
+     * @param info                  the roster and TSS information
+     */
+    void updateTssStatus(final boolean isStakePeriodBoundary, final Instant consensusNow, final RosterAndTssInfo info) {
+        final var statusChange = new StatusChange(isStakePeriodBoundary, consensusNow, info);
+        this.tssStatus = statusChange.computeNewStatus();
     }
 
     @VisibleForTesting
     public TssKeysAccessor getTssKeysAccessor() {
         return tssKeysAccessor;
     }
+
+    /**
+     * A class to manage the status change of the TSS.
+     * It computes the new status based on the old status and the current state of the system.
+     * If needed, it schedules work to generate TSS messages and votes.
+     */
+    public class StatusChange {
+        private TssKeyingStatus newKeyingStatus;
+        private RosterToKey newRosterToKey;
+        private Bytes newLedgerId;
+        private final boolean isStakePeriodBoundary;
+        private final Instant consensusNow;
+        private final RosterAndTssInfo info;
+
+        public StatusChange(
+                final boolean isStakePeriodBoundary, final Instant consensusNow, final RosterAndTssInfo info) {
+            this.isStakePeriodBoundary = isStakePeriodBoundary;
+            this.info = info;
+            this.newKeyingStatus = tssStatus.tssKeyingStatus();
+            this.newRosterToKey = tssStatus.rosterToKey();
+            this.newLedgerId = tssStatus.ledgerId();
+            this.consensusNow = consensusNow;
+        }
+
+        /**
+         * Computes the new status based on the old status and the current state of the system.
+         * If needed, it schedules work to generate TSS messages and votes.
+         *
+         * @return the new status
+         */
+        public TssStatus computeNewStatus() {
+            switch (tssStatus.rosterToKey()) {
+                case NONE -> {
+                    if (isStakePeriodBoundary) {
+                        newRosterToKey = CANDIDATE_ROSTER;
+                        newKeyingStatus = TssKeyingStatus.WAITING_FOR_THRESHOLD_TSS_MESSAGES;
```

Comment:
As above; since keying depends on having encryption keys for the **target** roster, it's possible that with _enough_ new members in the candidate roster, we have to return to `WAITING_FOR_ENCRYPTION_KEYS`.

Reply:
Makes sense. Changed it. Thanks